Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -202,24 +202,57 @@ instance : Neg (AdicCompletion I M) where
 instance : Sub (AdicCompletion I M) where
   sub x y := ⟨x.val - y.val, by simp [x.property, y.property]⟩
 
-instance : SMul ℕ (AdicCompletion I M) where
-  smul n x := ⟨n • x.val, by simp [x.property]⟩
+instance instSMul [SMul S R] [SMul S M] [IsScalarTower S R M] : SMul S (AdicCompletion I M) where
+  smul r x := ⟨r • x.val, by simp [x.property]⟩
+
+@[simp, norm_cast] lemma val_zero : (0 : AdicCompletion I M).val = 0 := rfl
+
+lemma val_zero_apply (n : ℕ) : (0 : AdicCompletion I M).val n = 0 := rfl
+
+variable {I M}
+
+@[simp, norm_cast] lemma val_add (f g : AdicCompletion I M) : (f + g).val = f.val + g.val := rfl
+@[simp, norm_cast] lemma val_sub (f g : AdicCompletion I M) : (f - g).val = f.val - g.val := rfl
+@[simp, norm_cast] lemma val_neg (f : AdicCompletion I M) : (-f).val = -f.val := rfl
 
-instance : SMul ℤ (AdicCompletion I M) where
-  smul n x := ⟨n • x.val, by simp [x.property]⟩
+lemma val_add_apply (f g : AdicCompletion I M) (n : ℕ) : (f + g).val n = f.val n + g.val n := rfl
+lemma val_sub_apply (f g : AdicCompletion I M) (n : ℕ) : (f - g).val n = f.val n - g.val n := rfl
+lemma val_neg_apply (f : AdicCompletion I M) (n : ℕ) : (-f).val n = -f.val n := rfl
+
+/- No `simp` attribute, since it causes `simp` unification timeouts when considering
+the `Module (AdicCompletion I R) (AdicCompletion I M)` instance (see `AdicCompletion/Algebra`). -/
+@[norm_cast]
+lemma val_smul [SMul S R] [SMul S M] [IsScalarTower S R M] (r : S) (f : AdicCompletion I M) :
```

Comment:
I think this lemma and the next would be easier to read if you named the variable `(s : S)` not `(r : S)` because there is also an `R` here.